File: python/questions/dailyTemperatures.py
from typing import List

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        lt = len(temperatures)
        st = []
        ans = [0] * lt
        for p in range(lt):
            ct = temperatures[p]
            while len(st) and ct>st[-1][0]:
                _, pp = st.pop()
                ans[pp] = p - pp
            st.append(([ct, p]))
        return ans

File: cpp/questions/dailyTemperatures.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int ts = temperatures.size();
        std::vector<int> ans(ts, 0);
        std::stack<std::pair<int, int>> st;
        for (int p=0; p<ts; ++p) {
            const int ct = temperatures[p];
            while (!st.empty() && ct>st.top().first) {
                ans[st.top().second] = p - st.top().second;
                st.pop();
            }
            st.emplace(ct, p);
        }
        return ans;
    }
};
